apps/vc-api/src/vc-api/exchanges/exchange.service.ts

```diff
@@ -107,6 +107,13 @@ export class ExchangeService {
     }
     const transaction = transactionQuery.transaction;
 
+    if (
+      transaction.presentationSubmission &&
+      transaction.presentationSubmission.vp.holder !== verifiablePresentation.holder
+    ) {
+      throw new ForbiddenException('DID does not match the DID that initially submitted the presentation');
+    }
```

Comment:
IMO, it is better if this logic is in entity (`transaction.processPresentation`). To me, this is domain logic and so it should in the domain object. In theory, maybe we could extract the entities one day as a library and import them into NestJs for example (in a kind of onion architecture approach). What are your thoughts on this @artursudnik ?

Reply:
Yes, I agree. I have a strong need to throw ForbiddenException :) but it does not make sense to do this from within the entity.
I need to think about how to refactor this to still have this exception thrown by the service.

Reply:
My initial thought on how to handle it is to have a list of domain errors that the Nest code (service or controller) can map to the appropriate HTTP response. But maybe you have a better approach as well

Reply:
I agree that this is a good approach. I have implemented something like this in the Zero project, where Prisma ORM throws exceptions with code field and [NestJS converts](https://github.com/energywebfoundation/zero/blob/develop/apps/zero-api/src/exception-filters/PrismaClientExceptionFilter.ts) its value into response errors.

Reply:
I am moving this logic to the entity for now and I think it will throw ForbiddenException until we have an idea of how to make error handling consistent. Now, this code is an integral part of the service, so this is not an issue. Anyway, it is not compatible with the `error` field I see is returned by many methods. If we decide to have a list of domain exceptions thrown, it will reduce the amount of code needed for error handling and will reduce the probability of mistakes made while writing code by not checking if the `error` field is empty. Exception always interrupts execution of the code if not handled.

Reply:
Sounds good!

Reply:
I am thinking about something like this:
```typescript
export class EwSsiException extends Error{
  constructor(message) {
    super(message);
    this.initName();
  }

  public initName(): void {
    this.name = this.constructor.name;
  }
}

export class TransactionEntityException extends EwSsiException {}

export class TransactionDidForbiddenException extends TransactionEntityException {}
```
and then:
```typescript
throw new TransactionDidForbiddenException('DID does not match the DID that initially submitted the presentation');
```
Then we can have the NestJS exception filter handle this.
or just:
```typescript
if (err instanceof TransactionDidForbiddenException) {
  throw new ForbiddenException(err.message);
}
```
The trickiest part is to design a hierarchy of these exceptions well to have it granular enough but not too much...

Reply:
This approach of using Exception classes sounds good to me. I think the hierarchy is not so important to start, maybe having them grouped by entity is good enough. If another grouping/hierarchy emerges in the future, we can refactor to that